Hw17 spring mvc

securityboxes opened the pull request:
Spring MVC

ihavedone committed:
en: Spring MVC
ru: Spring MVC

Commit:
add library-front (react).

stvort commented on library-front/src/Application.js:
```diff
@@ -0,0 +1,42 @@
+                    </Route>
+                    <Route exact path="/edit/:id" component={BookForm}>
+                    </Route>
+                    )}/>
```
Возможно `)}/>` тут лишнее?

ihavedone replied:
Действительно, осталось от рефакторинга

stvort commented on library-front/src/BookForm.js:
```diff
@@ -0,0 +1,135 @@
+
+class BookForm extends React.Component {
+
+    constructor(props) {
+        super(props);
+        this.state = {
+            book_caption: "",
+            book_id: null,
+            book_authors: "",
+            book_genres: "",
+            redirect: false
+        }
+        this.handleChangeCaption = this.handleChangeCaption.bind(this);
+        this.handleChangeAuthors = this.handleChangeAuthors.bind(this);
+        this.handleChangeGenres = this.handleChangeGenres.bind(this);
+        this.saveHandler = this.saveHandler.bind(this);
+    }
```
В реакте существует практика, когда состоянием обладает один родительский компонент, а все что нужно дочерним спускается им в пропсах. В т.ч. методы для изменения состояния.. Т.о. классом будет только этот компонент с состоянием, а остальных можно писать функциями. Плюс появляется возможность сохранить/восстановить состояние целиком. В качестве альтернативы этому подходу используют redux

ihavedone replied:
Стейтлесс перевел в функции, пару стейтфул оставил (с целью экономии времени пока), идею понял. Благодаря этому курсу в будущем схожу на курсы по реакту)

stvort commented on library-front/src/Book.js:
```diff
@@ -0,0 +1,23 @@
+import Author from "./Author";
+import Genre from "./Genre";
+
+class Book extends React.Component {
```
В общем чем больше компонентов будут функциями тем лучше)

stvort commented on library-front/src/BookForm.js:
```diff
@@ -0,0 +1,135 @@
+        }
+    }
+
+    getServerData(id) {
+        fetch('/api/book/' + id)
+            .then(result => result.json())
+            .then((result) => {
+                let authors = ""
+                result.authors.forEach(a => authors += a.name + ", ")
+                authors = authors.slice(0, authors.length - 2)
+                let genres = ""
+                result.genres.forEach(g => genres += g.name + ", ")
+                genres = genres.slice(0, genres.length - 2)
+                this.setState({
+                    book_caption: result.caption,
+                    book_id: result.id,
+                    book_authors: authors,
+                    book_genres: genres
+                })
+            })
+    }
```
Методы для работы с апи лучше убрать в сервисный модуль

stvort commented on library-front/src/BookForm.js:
```diff
@@ -0,0 +1,135 @@
+            <div className="grid">
+                <input type="hidden" id="book_id" onChange={() => {
+                }} value={this.state.book_id}/>
+                <div className="row"><span className="cell3">Caption:</span>
+                    <span className="cell7"><input type="text" className="textField" id="book_caption"
+                                                   onChange={this.handleChangeCaption}
+                                                   placeholder="Caption of the book" value={this.state.book_caption}/>
+                    </span>
+                </div>
```
Возможно это тоже компонент)

stvort commented on library-front/src/BookList.js:
```diff
@@ -0,0 +1,62 @@
+        return (
+            <div>
+                {this.state.books.map((book, index) => {
+                    return <div className="bookBlock">
+                        <Book key={book.id} id={book.id}
+                              caption={book.caption}
+                              authors={book.authors}
+                              genres={book.genres}
+                        />
+                        <button onClick={() => this.deleteBook(book.id)}>Delete</button>
+                        <Link to={"/edit/" + book.id}>
+                            <button>Edit</button>
+                        </Link>
+                    </div>
```
И это)

ihavedone replied:
Так реализовал, потому что не подумал о том, чтобы передать обработчик как ссылку на функцию в дочерний объект. Переделал, получился один компонент Book в том виде, в котором задумал его заранее :)

Commit:
en: front refactoring. Move API to the dedicated service, transform stateless components to functions, rebuild Book class
ru: рефакторинг фронтовой части. Общение с API - в отдельный сервисный компонент, компоненты без состояния разжалованы до функций, оптимизирован компонент Book
